[macOS] All Writing Tools menu items are enabled in app menus for empty editable content

pxlcoder opened the pull request:
#### 0dab99e9279ea0df37f64c3613c6bb7a637fc2d8
<pre>
[macOS] All Writing Tools menu items are enabled in app menus for empty editable content
<a href="https://bugs.webkit.org/show_bug.cgi?id=286310">https://bugs.webkit.org/show_bug.cgi?id=286310</a>
<a href="https://rdar.apple.com/143332082">rdar://143332082</a>

Reviewed by Wenson Hsieh.

284869@main fixed an issue where all Writing Tools menu items were enabled
in the context menu for empty editable content.

Implement the same policy for the menu items in the app menu:

1. &quot;Show Writing Tools&quot; should always be enabled.
2. &quot;Compose&quot; should only be enabled in editable content.
3. Other tools should be enabled if the content is non-editable, or if the editable content is non-empty.

This is achieved by implementing by checking for `showWritingTools:` in
`validateUserInterfaceItem:`, and sharing the code introduced in 284869@main.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
(WebKit::WebContextMenuProxyMac::getContextMenuItem):
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::shouldEnableWritingToolsRequestedTool const):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::validateUserInterfaceItem):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(TEST(WritingTools, AppMenuNonEditable)):
(TEST(WritingTools, AppMenuEditable)):
(TEST(WritingTools, AppMenuEditableEmpty)):

Canonical link: <a href="https://commits.webkit.org/289243@main">https://commits.webkit.org/289243@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a596231e181b11c1160209bdfd71cb937359c5bb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85925 "") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5546 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/40293 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90938 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36836 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87970 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/5785 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/13550 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66682 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24478 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88928 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/4423 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77938 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46972 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4276 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/32296 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35919 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74908 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/33085 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/92729 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/13174 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9660 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/75475 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/13388 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73789 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/74637 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18809 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/17247 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/6048 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13390 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/13207 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/18555 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12980 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/16413 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/14767 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
```diff
@@ -919,6 +920,28 @@ static inline bool expectsLegacyImplicitRubberBandControl()
+    bool allowProofreadingAndRewritingTools = true;
+
+    auto& editorState = this->editorState();
+    editorStateIsContentEditable = editorState.isContentEditable;
```
the variable can be declared and defined on this line

pxlcoder replied:
Done.

rr-codes commented on Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
```diff
@@ -919,6 +920,28 @@ static inline bool expectsLegacyImplicitRubberBandControl()
+    if (editorStateIsContentEditable)
+        allowProofreadingAndRewritingTools = editorState.hasPostLayoutData() && !editorState.postLayoutData->paragraphContextForCandidateRequest.isEmpty();
+
+    if (requestedTool == WTRequestedToolCompose)
```
this if can be moved above to the preceding if to possibly avoid an extra if

pxlcoder replied:
Done. And also got rid of the `allowProofreadingAndRewritingTools` variable.

aprotyas commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
```diff
@@ -3271,6 +3272,66 @@ InstanceMethodSwizzler swizzler(PAL::getWTWritingToolsClass(), @selector(showToo
+    RetainPtr writingToolsMenuItem = [NSMenuItem standardWritingToolsMenuItem];
+    EXPECT_NOT_NULL(writingToolsMenuItem.get());
+
+    RetainPtr<NSArray<NSMenuItem *>> items = [writingToolsMenuItem submenu].itemArray;
```
CTAD?
```suggestion
    RetainPtr items = [writingToolsMenuItem submenu].itemArray;
```

whsieh replied:
IIRC, template type deduction doesn't work well when using ObjC properties like `….itemArray`. It may work if you change it to `[… itemArray]`, however.

pxlcoder replied:
Wenson's suggestion works.

aprotyas replied:
🎉

pxlcoder committed:
[macOS] All Writing Tools menu items are enabled in app menus for empty editable content
https://bugs.webkit.org/show_bug.cgi?id=286310
rdar://143332082

Reviewed by Wenson Hsieh.

284869@main fixed an issue where all Writing Tools menu items were enabled
in the context menu for empty editable content.

Implement the same policy for the menu items in the app menu:

1. "Show Writing Tools" should always be enabled.
2. "Compose" should only be enabled in editable content.
3. Other tools should be enabled if the content is non-editable, or if the editable content is non-empty.

This is achieved by implementing by checking for `showWritingTools:` in
`validateUserInterfaceItem:`, and sharing the code introduced in 284869@main.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
(WebKit::WebContextMenuProxyMac::getContextMenuItem):
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::shouldEnableWritingToolsRequestedTool const):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::validateUserInterfaceItem):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(TEST(WritingTools, AppMenuNonEditable)):
(TEST(WritingTools, AppMenuEditable)):
(TEST(WritingTools, AppMenuEditableEmpty)):

Canonical link: https://commits.webkit.org/289243@main
